### PR TITLE
Hotfix/p9 milc

### PR DESCRIPTION
**Pull request by maddyscientist**
Some fixes and cleanup in this pull
* When allocating mapped memory, use `cudaAllocHost` instead of `malloc` and `cudaHostRegister` (work around for issue on POWER9 systems)
* Improved test coverage of `unitarizeLinkTest` to use mapped memory matching what the actual interface does
* Fixed bug in link unitarization (only first parity was being unitarized) and some performance optimization
* Fix bug in HISQ force outer product when including a non-zero Naik term in the action
* In MILC interface, when using QMP communication, do not supply a rank-to-coordinate functor, allowing QMP to specify the logical topology
* Remove legacy some `volatile` declarations

### CODE REVIEW

File: lib/malloc.cpp [hunk 1]
```diff
@@ -269,6 +269,7 @@ namespace quda {
     return ptr;
   }
 
+#define HOST_ALLOC // this needs to be set presently on P9
```
Comment (mathiaswagner) [inline lib/malloc.cpp]: should this be defined by cmake?

Comment (maddyscientist) [reply lib/malloc.cpp]: I don't think so.  Making it a cmake parameter means a global rebuild when changing it, where as for debugging purposes you really want to be able to make a local change and quick rebuild.  This isn't something we ever need to change, but I left the old code in there for future testing.

Comment (mathiaswagner) [reply lib/malloc.cpp]: Ok. Was not sure whether this should be enabled only on Power.

File: lib/malloc.cpp [hunk 2]
```diff
@@ -363,26 +373,35 @@ namespace quda {
...
+      free(ptr);
     } else if (alloc[MAPPED].count(ptr)) {
+#ifdef HOST_ALLOC
+      cudaError_t err = cudaFreeHost(ptr);
+      if (err != cudaSuccess) {
+	printfQuda("ERROR: Failed to free host memory (%s:%d in %s())\n", file, line, func);
+	errorQuda("Aborting");
+      }
+#else
       cudaError_t err = cudaHostUnregister(ptr);
```
Comment (mathiaswagner) [inline lib/malloc.cpp]: do we miss a free(ptr) here if HOST_ALLOC is not defined?

Comment (maddyscientist) [reply lib/malloc.cpp]: Good catch.  Will push a fix

**Pull request merged**